File: analytics.py
import os
import logging
from datetime import datetime
from typing import Optional, Dict, List, Tuple

from supabase import create_client, Client

logger = logging.getLogger(__name__)


class Analytics:
    """Analytics tracking for the Telegram bot using Supabase."""
    
    def __init__(self):
        """Initialize analytics with Supabase connection."""
        supabase_url = os.getenv("SUPABASE_URL")
        supabase_key = os.getenv("SUPABASE_KEY")
        
        if not supabase_url or not supabase_key:
            logger.warning("Supabase credentials not found. Analytics will be disabled.")
            self.client = None
            return
        
        self.client: Client = create_client(supabase_url, supabase_key)
        logger.info("Supabase analytics initialized")
    
    def _is_enabled(self) -> bool:
        """Check if analytics is enabled."""
        return self.client is not None
    
    def track_user(self, user_id: int, username: Optional[str] = None, 
                   first_name: Optional[str] = None, last_name: Optional[str] = None,
                   language_code: Optional[str] = None):
        """Track or update user information."""
        if not self._is_enabled():
            return
        
        try:
            now = datetime.now().isoformat()
            
            # Check if user exists
            result = self.client.table('bot_users').select('user_id').eq('user_id', user_id).execute()
            
            if result.data:
                # Update existing user
                self.client.table('bot_users').update({
                    'username': username,
                    'first_name': first_name,
                    'last_name': last_name,
                    'language_code': language_code,
                    'last_seen': now
                }).eq('user_id', user_id).execute()
            else:
                # Insert new user
                self.client.table('bot_users').insert({
                    'user_id': user_id,
                    'username': username,
                    'first_name': first_name,
                    'last_name': last_name,
                    'language_code': language_code,
                    'first_seen': now,
                    'last_seen': now
                }).execute()
        except Exception as e:
            logger.error(f"Error tracking user: {e}")
    
    def track_event(self, user_id: int, event_type: str, 
                   media_type: Optional[str] = None, chat_type: Optional[str] = None):
        """Track an event (transcription, summary, etc.)."""
        if not self._is_enabled():
            return
        
        try:
            self.client.table('bot_events').insert({
                'user_id': user_id,
                'event_type': event_type,
                'media_type': media_type,
                'chat_type': chat_type
            }).execute()
        except Exception as e:
            logger.error(f"Error tracking event: {e}")
    
    def get_total_stats(self) -> Dict[str, int]:
        """Get overall statistics."""
        if not self._is_enabled():
            return {'total_transcriptions': 0, 'total_summaries': 0, 'total_users': 0, 'total_events': 0}
        
        try:
            # Count transcriptions
            trans_result = self.client.table('bot_events').select('id', count='exact').eq('event_type', 'transcription').execute()
            total_transcriptions = trans_result.count or 0
            
            # Count summaries
            sum_result = self.client.table('bot_events').select('id', count='exact').eq('event_type', 'summary').execute()
            total_summaries = sum_result.count or 0
            
            # Count users
            users_result = self.client.table('bot_users').select('user_id', count='exact').execute()
            total_users = users_result.count or 0
            
            # Count all events
            events_result = self.client.table('bot_events').select('id', count='exact').execute()
            total_events = events_result.count or 0
            
            return {
                'total_transcriptions': total_transcriptions,
                'total_summaries': total_summaries,
                'total_users': total_users,
                'total_events': total_events
            }
        except Exception as e:
            logger.error(f"Error getting total stats: {e}")
            return {'total_transcriptions': 0, 'total_summaries': 0, 'total_users': 0, 'total_events': 0}
    
    def get_media_type_stats(self) -> Dict[str, int]:
        """Get statistics by media type."""
        if not self._is_enabled():
            return {}
        
        try:
            result = self.client.table('bot_events').select('media_type').eq('event_type', 'transcription').not_.is_('media_type', 'null').execute()
            
            stats = {}
            for row in result.data:
                media_type = row['media_type']
                stats[media_type] = stats.get(media_type, 0) + 1
            return stats
        except Exception as e:
            logger.error(f"Error getting media type stats: {e}")
            return {}
    
    def get_top_users(self, limit: int = 10) -> List[Tuple[int, str, int]]:
        """Get top users by request count."""
        if not self._is_enabled():
            return []
        
        try:
            # Get all events with user info
            result = self.client.table('bot_events').select('user_id').execute()
            
            # Count events per user
            user_counts = {}
            for row in result.data:
                user_id = row['user_id']
                user_counts[user_id] = user_counts.get(user_id, 0) + 1
            
            # Get user details
            users_result = self.client.table('bot_users').select('user_id, username').execute()
            user_names = {u['user_id']: u['username'] for u in users_result.data}
            
            # Combine and sort
            top_users = [(user_id, user_names.get(user_id), count) 
                        for user_id, count in user_counts.items()]
            top_users.sort(key=lambda x: x[2], reverse=True)
            
            return top_users[:limit]
        except Exception as e:
            logger.error(f"Error getting top users: {e}")
            return []
    
    def get_user_stats(self, user_id: int) -> Dict[str, int]:
        """Get statistics for a specific user."""
        if not self._is_enabled():
            return {'transcriptions': 0, 'summaries': 0, 'total': 0}
        
        try:
            result = self.client.table('bot_events').select('event_type').eq('user_id', user_id).execute()
            
            stats = {'transcription': 0, 'summary': 0}
            for row in result.data:
                event_type = row['event_type']
                if event_type in stats:
                    stats[event_type] += 1
            
            return {
                'transcriptions': stats.get('transcription', 0),
                'summaries': stats.get('summary', 0),
                'total': sum(stats.values())
            }
        except Exception as e:
            logger.error(f"Error getting user stats: {e}")
            return {'transcriptions': 0, 'summaries': 0, 'total': 0}
    
    def get_language_distribution(self) -> Dict[str, int]:
        """Get distribution of user languages."""
        if not self._is_enabled():
            return {}
        
        try:
            result = self.client.table('bot_users').select('language_code').not_.is_('language_code', 'null').execute()
            
            stats = {}
            for row in result.data:
                lang = row['language_code']
                stats[lang] = stats.get(lang, 0) + 1
            return dict(sorted(stats.items(), key=lambda x: x[1], reverse=True))
        except Exception as e:
            logger.error(f"Error getting language distribution: {e}")
            return {}
    
    def get_chat_type_stats(self) -> Dict[str, int]:
        """Get statistics by chat type (private vs group)."""
        if not self._is_enabled():
            return {}
        
        try:
            result = self.client.table('bot_events').select('chat_type').not_.is_('chat_type', 'null').execute()
            
            stats = {}
            for row in result.data:
                chat_type = row['chat_type']
                stats[chat_type] = stats.get(chat_type, 0) + 1
            return stats
        except Exception as e:
            logger.error(f"Error getting chat type stats: {e}")
            return {}
    
    def get_daily_stats(self, days: int = 7) -> List[Tuple[str, int]]:
        """Get statistics for the last N days."""
        if not self._is_enabled():
            return []
        
        try:
            from datetime import timedelta
            cutoff = (datetime.now() - timedelta(days=days)).isoformat()
            
            result = self.client.table('bot_events').select('created_at').gte('created_at', cutoff).execute()
            
            # Group by day
            daily_counts = {}
            for row in result.data:
                day = row['created_at'][:10]  # Extract YYYY-MM-DD
                daily_counts[day] = daily_counts.get(day, 0) + 1
            
            return sorted(daily_counts.items(), reverse=True)
        except Exception as e:
            logger.error(f"Error getting daily stats: {e}")
            return []
    
    def get_monthly_stats(self, months: int = 6) -> List[Tuple[str, int]]:
        """Get statistics for the last N months."""
        if not self._is_enabled():
            return []
        
        try:
            from datetime import timedelta
            cutoff = (datetime.now() - timedelta(days=months * 30)).isoformat()
            
            result = self.client.table('bot_events').select('created_at').gte('created_at', cutoff).execute()
            
            # Group by month
            monthly_counts = {}
            for row in result.data:
                month = row['created_at'][:7]  # Extract YYYY-MM
                monthly_counts[month] = monthly_counts.get(month, 0) + 1
            
            return sorted(monthly_counts.items(), reverse=True)
        except Exception as e:
            logger.error(f"Error getting monthly stats: {e}")
            return []
    
    def get_yearly_stats(self) -> List[Tuple[str, int]]:
        """Get statistics by year."""
        if not self._is_enabled():
            return []
        
        try:
            result = self.client.table('bot_events').select('created_at').execute()
            
            # Group by year
            yearly_counts = {}
            for row in result.data:
                year = row['created_at'][:4]  # Extract YYYY
                yearly_counts[year] = yearly_counts.get(year, 0) + 1
            
            return sorted(yearly_counts.items(), reverse=True)
        except Exception as e:
            logger.error(f"Error getting yearly stats: {e}")
            return []
    
    def get_hourly_distribution(self) -> Dict[int, int]:
        """Get distribution of requests by hour of day (0-23)."""
        if not self._is_enabled():
            return {}
        
        try:
            result = self.client.table('bot_events').select('created_at').execute()
            
            # Group by hour
            hourly_counts = {}
            for row in result.data:
                # Parse ISO timestamp and extract hour
                hour = int(row['created_at'][11:13])
                hourly_counts[hour] = hourly_counts.get(hour, 0) + 1
            
            return dict(sorted(hourly_counts.items()))
        except Exception as e:
            logger.error(f"Error getting hourly distribution: {e}")
            return {}

File: supabase/functions/telegram-webhook/analytics.ts
// Analytics module for tracking bot usage
// Converted from Python analytics.py
// Uses Supabase PostgreSQL instead of SQLite

import { SupabaseClient } from 'https://esm.sh/@supabase/supabase-js@2';
import { TotalStats, UserRecord, EventRecord } from './types.ts';

export class Analytics {
    private supabase: SupabaseClient;

    constructor(supabase: SupabaseClient) {
        this.supabase = supabase;
    }

    /**
     * Track or update user information
     */
    async trackUser(
        userId: number,
        username?: string,
        firstName?: string,
        lastName?: string,
        languageCode?: string
    ): Promise<void> {
        try {
            const now = new Date();

            // Check if user exists
            const { data: existingUser } = await this.supabase
                .from('users')
                .select('user_id')
                .eq('user_id', userId)
                .single();

            if (existingUser) {
                // Update existing user
                await this.supabase
                    .from('users')
                    .update({
                        username,
                        first_name: firstName,
                        last_name: lastName,
                        language_code: languageCode,
                        last_seen: now.toISOString()
                    })
                    .eq('user_id', userId);
            } else {
                // Insert new user
                await this.supabase
                    .from('users')
                    .insert({
                        user_id: userId,
                        username,
                        first_name: firstName,
                        last_name: lastName,
                        language_code: languageCode,
                        first_seen: now.toISOString(),
                        last_seen: now.toISOString()
                    });
            }
        } catch (error) {
            console.error('Error tracking user:', error);
        }
    }

    /**
     * Track an event (transcription, summary, etc.)
     */
    async trackEvent(
        userId: number,
        eventType: 'transcription' | 'summary',
        mediaType?: 'voice' | 'video_note' | 'audio' | 'video',
        chatType?: 'private' | 'group'
    ): Promise<void> {
        try {
            await this.supabase
                .from('events')
                .insert({
                    user_id: userId,
                    event_type: eventType,
                    media_type: mediaType,
                    chat_type: chatType,
                    timestamp: new Date().toISOString()
                });
        } catch (error) {
            console.error('Error tracking event:', error);
        }
    }

    /**
     * Get overall statistics
     */
    async getTotalStats(): Promise<TotalStats> {
        try {
            // Get event counts by type
            const { data: events } = await this.supabase
                .from('events')
                .select('event_type');

            const transcriptions = events?.filter(e => e.event_type === 'transcription').length || 0;
            const summaries = events?.filter(e => e.event_type === 'summary').length || 0;

            // Get unique users count
            const { count: userCount } = await this.supabase
                .from('users')
                .select('user_id', { count: 'exact', head: true });

            // Get total events count
            const { count: eventCount } = await this.supabase
                .from('events')
                .select('id', { count: 'exact', head: true });

            return {
                total_transcriptions: transcriptions,
                total_summaries: summaries,
                total_users: userCount || 0,
                total_events: eventCount || 0
            };
        } catch (error) {
            console.error('Error getting total stats:', error);
            return {
                total_transcriptions: 0,
                total_summaries: 0,
                total_users: 0,
                total_events: 0
            };
        }
    }

    /**
     * Get statistics by media type
     */
    async getMediaTypeStats(): Promise<Record<string, number>> {
        try {
            const { data } = await this.supabase
                .from('events')
                .select('media_type')
                .eq('event_type', 'transcription')
                .not('media_type', 'is', null);

            const stats: Record<string, number> = {};
            data?.forEach(row => {
                if (row.media_type) {
                    stats[row.media_type] = (stats[row.media_type] || 0) + 1;
                }
            });

            return stats;
        } catch (error) {
            console.error('Error getting media type stats:', error);
            return {};
        }
    }

    /**
     * Get top users by request count
     */
    async getTopUsers(limit: number = 10): Promise<Array<[number, string | null, number]>> {
        try {
            const { data } = await this.supabase
                .from('users')
                .select(`
          user_id,
          username,
          events:events(count)
        `)
                .order('events.count', { ascending: false })
                .limit(limit);

            if (!data) return [];

            // Transform to match Python format: [user_id, username, count]
            return data.map(row => [
                row.user_id,
                row.username || null,
                (row.events as any)?.length || 0
            ]);
        } catch (error) {
            console.error('Error getting top users:', error);
            return [];
        }
    }

    /**
     * Get distribution of user languages
     */
    async getLanguageDistribution(): Promise<Record<string, number>> {
        try {
            const { data } = await this.supabase
                .from('users')
                .select('language_code')
                .not('language_code', 'is', null);

            const stats: Record<string, number> = {};
            data?.forEach(row => {
                if (row.language_code) {
                    stats[row.language_code] = (stats[row.language_code] || 0) + 1;
                }
            });

            return stats;
        } catch (error) {
            console.error('Error getting language distribution:', error);
            return {};
        }
    }

    /**
     * Get statistics by chat type (private vs group)
     */
    async getChatTypeStats(): Promise<Record<string, number>> {
        try {
            const { data } = await this.supabase
                .from('events')
                .select('chat_type')
                .not('chat_type', 'is', null);

            const stats: Record<string, number> = {};
            data?.forEach(row => {
                if (row.chat_type) {
                    stats[row.chat_type] = (stats[row.chat_type] || 0) + 1;
                }
            });

            return stats;
        } catch (error) {
            console.error('Error getting chat type stats:', error);
            return {};
        }
    }

    /**
     * Get statistics for the last N days
     */
    async getDailyStats(days: number = 7): Promise<Array<[string, number]>> {
        try {
            const daysAgo = new Date();
            daysAgo.setDate(daysAgo.getDate() - days);

            const { data } = await this.supabase
                .from('events')
                .select('timestamp')
                .gte('timestamp', daysAgo.toISOString());

            // Group by date
            const stats: Record<string, number> = {};
            data?.forEach(row => {
                const date = new Date(row.timestamp).toISOString().split('T')[0];
                stats[date] = (stats[date] || 0) + 1;
            });

            // Convert to array and sort by date descending
            return Object.entries(stats).sort((a, b) => b[0].localeCompare(a[0]));
        } catch (error) {
            console.error('Error getting daily stats:', error);
            return [];
        }
    }

    /**
     * Get statistics for the last N months
     */
    async getMonthlyStats(months: number = 6): Promise<Array<[string, number]>> {
        try {
            const monthsAgo = new Date();
            monthsAgo.setMonth(monthsAgo.getMonth() - months);

            const { data } = await this.supabase
                .from('events')
                .select('timestamp')
                .gte('timestamp', monthsAgo.toISOString());

            // Group by year-month
            const stats: Record<string, number> = {};
            data?.forEach(row => {
                const yearMonth = new Date(row.timestamp).toISOString().substring(0, 7); // YYYY-MM
                stats[yearMonth] = (stats[yearMonth] || 0) + 1;
            });

            // Convert to array and sort by month descending
            return Object.entries(stats).sort((a, b) => b[0].localeCompare(a[0]));
        } catch (error) {
            console.error('Error getting monthly stats:', error);
            return [];
        }
    }

    /**
     * Get statistics by year
     */
    async getYearlyStats(): Promise<Array<[string, number]>> {
        try {
            const { data } = await this.supabase
                .from('events')
                .select('timestamp');

            // Group by year
            const stats: Record<string, number> = {};
            data?.forEach(row => {
                const year = new Date(row.timestamp).getFullYear().toString();
                stats[year] = (stats[year] || 0) + 1;
            });

            // Convert to array and sort by year descending
            return Object.entries(stats).sort((a, b) => b[0].localeCompare(a[0]));
        } catch (error) {
            console.error('Error getting yearly stats:', error);
            return [];
        }
    }

    /**
     * Get distribution of requests by hour of day (0-23)
     */
    async getHourlyDistribution(): Promise<Record<number, number>> {
        try {
            const { data } = await this.supabase
                .from('events')
                .select('timestamp');

            // Group by hour
            const stats: Record<number, number> = {};
            data?.forEach(row => {
                const hour = new Date(row.timestamp).getUTCHours();
                stats[hour] = (stats[hour] || 0) + 1;
            });

            return stats;
        } catch (error) {
            console.error('Error getting hourly distribution:', error);
            return {};
        }
    }
}
